x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java

```diff
@@ -32,6 +32,91 @@
+ * interact with in some way. This includes fully representable types (e.g.
+ * {@link DataType#LONG}, numeric types which we promote (e.g. {@link DataType#SHORT})
+ * or fold into other types (e.g. {@link DataType#DATE_PERIOD}) early in the
+ * processing pipeline, types for internal use
+ * cases (e.g. {@link DataType#PARTIAL_AGG}), and types which the language
+ * doesn't support, but require special handling anyway (e.g.
+ * {@link DataType#OBJECT})
+ *
+ * <h2>Process for adding a new data type</h2>
+ * Note: it is not expected that all the following steps be done in a single PR.
+ * Use capabilities to gate tests as you go, and use as many PRs as you think
+ * appropriate. New data types are complex, and smaller PRs will make reviews
+ * easier.
+ * <ul>
+ *     <li>Create a new feature flag for the type in {@link EsqlCorePlugin}. We
+ *     recommend developing the data type over a series of smaller PRs behind
+ *     a feature flag; even for relatively simple data types.</li>
+ *     <li>Add a capability to EsqlCapabilities related to the new type, and
+ *     gated by the feature flag you just created. Again, using the feature
+ *     flag is preferred over snapshot-only. As development progresses, you may
+ *     need to add more capabilities related to the new type, e.g. for
+ *     supporting specific functions. This is fine, and expected.</li>
+ *     <li>Create a new CSV test file for the new type. Depending on the type,
+ *     you may also need to create a new sample data file if none of the
+ *     existing sample data files have fields of that type. See
+ *     CsvTestDataLoader for creating a new data set.</li>
+ *     <li>In the new CSV test file, start adding basic functionality tests.
+ *     These should include reading and returning values, both from indexed data
+ *     and from the ROW command.  It should also include functions that support
+ *     "every" type, such as Case or MvFirst.</li>
+ *     <li>Add the new type to the CsvTestUtils#Type enum, if it isn't already
+ *     there. You may also need to modify CsvAssert to support reading values
+ *     of the new type.</li>
+ *     <li>At this point, the CSV tests should fail with a sensible ES|QL error
+ *     message. Make sure they're failing in ES|QL, not in the test
+ *     framework.</li>
+ *     <li>Add the new data type to this enum. This will cause a bunch of
+ *     compile errors for switch statements throughout the code.  Resolve those
+ *     as appropriate. That is the main way in which the new type will be tied
+ *     into the framework.</li>
+ *     <li>Add the new type to the {@link DataType#UNDER_CONSTRUCTION}
+ *     collection. This is used by the test framework to disable some checks
+ *     around how functions report their supported types, which would otherwise
+ *     generate a lot of noise while the type is still in development.</li>
+ *     <li>Add typed data generators to TestCaseSupplier, and make sure all
```

Comment:
I think your first PR should end somewhere around here, right? Maybe a note about that?

Reply:
I was trying to avoid being prescriptive about how to break up PRs.

Reply:
Yeah. I agree on not being prescriptive. But maybe "this is a good place to stop the first PR" or, I dunno. Something. Or not. It's fine.